components/brave_rewards/resources/extension/brave_rewards/background/publisherVisit.ts

```diff
@@ -0,0 +1,240 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+interface GreaselionErrorResponse {
+  errorMessage: string
+}
+
+interface SavePublisherVisitResponse {
+  url: string
+  publisherKey: string
+  publisherName: string
+  mediaKey?: string
+  favIconUrl?: string
+}
+
+interface GreaselionResponse {
+  msgType: 'GreaselionError' | 'SavePublisherVisit'
+  mediaType: string
+  data: GreaselionErrorResponse | SavePublisherVisitResponse | null
+}
+
+interface MediaDurationHandlerRegistrationRequest {
+  urlRegex: string
+}
+
+interface MediaDurationMetadataResponse {
+  mediaKey: string
+  duration: number
+}
+
+// Media duration handler registration sentinel
+interface MediaDurationHandlerRegistration {
+  registered: boolean
+}
+
+// Maps a tabId to a registration sentinel
+interface MediaDurationHandlerRegistrationDictionary {
+  [index: number]: MediaDurationHandlerRegistration
+}
+
+// Maps a media key to a specific publisher
+interface PublisherKeyDictionary {
+  [index: string]: string
+}
+
+// Maps a tabId to a specific Greaselion port
+interface GreaselionPortDictionary {
+  [index: number]: chrome.runtime.Port
+}
+
+let greaselionPorts: GreaselionPortDictionary = {}
+
+let mediaDurationHandlerRegistrations: MediaDurationHandlerRegistrationDictionary = {}
+
+let publisherKeys: PublisherKeyDictionary = {}
+
+const handleGreaselionErrorResponse = (tabId: number, mediaType: string, data: GreaselionErrorResponse) => {
+  console.error(`Greaselion error: ${data.errorMessage}`)
+}
+
+const handleMediaDurationMetadataResponse = (tabId: number, mediaType: string, data: MediaDurationMetadataResponse) => {
+  const publisherKey = publisherKeys[data.mediaKey]
+  if (!publisherKey) {
+    console.error(`Failed to handle media duration metadata response: missing publisher key for media key ${data.mediaKey}`)
+    return
+  }
+
+  chrome.braveRewards.getPublisherInfo(publisherKey, (result: number, info?: RewardsExtension.Publisher) => {
+    console.debug(`getPublisherInfo: result=${result}`)
+
+    if (result === 0 && info) {
+      console.info('Updating media duration:')
```

Comment:
please remove this before merging as we can't log things that exposes browsing history

Reply:
Removed.